### PR TITLE
helm: Enable persistence in custom mode

### PR DESCRIPTION
**Pull request by pierreyves-lebrun**
Users might want to persist data while providing a custom config file, this PR enables persistence in that scenario

### CODE REVIEW

**Comment by webvictim** [comment]
@pierreyves-lebrun Thanks for the PR! Could you please take a look at https://github.com/gravitational/teleport/blob/master/examples/chart/CONTRIBUTING.md and make sure to do everything on the list?

File: examples/chart/teleport-cluster/Chart.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 name: teleport-cluster
 apiVersion: v2
-version: "10.0.0-dev"
+version: "10.0.1-dev"
```
Comment (webvictim) [inline examples/chart/teleport-cluster/Chart.yaml:3]:
```suggestion
```

You can remove this edit, the version is automatically updated when we push releases.

File: examples/chart/teleport-cluster/templates/deployment.yaml [hunk 2]
```diff
@@ -190,9 +190,9 @@ spec:
         configMap:
           name: {{ .Release.Name }}
       - name: "data"
-        {{- if eq .Values.chartMode "standalone" }}
+        {{- if or (eq .Values.chartMode "standalone") (eq .Values.chartMode "custom") }}
         persistentVolumeClaim:
-          claimName: {{ if .Values.standalone.existingClaimName }}{{ .Values.standalone.existingClaimName }}{{ else }}{{ .Release.Name }}{{ end }}
+          claimName: {{ if .Values.persistence.existingClaimName }}{{ .Values.persistence.existingClaimName }}{{ else }}{{ .Release.Name }}{{ end }}
```
Comment (webvictim) [inline examples/chart/teleport-cluster/templates/deployment.yaml:195]: Another thing we'd also need to do is support backward compatibility for this, i.e. someone who still has `standalone.existingClaimName` set in their values shouldn't have their chart broken by this change. I'd suggest using `coalesce` to pick the first non-empty value in a similar way to https://github.com/gravitational/teleport/blob/master/examples/chart/teleport-cluster/templates/config.yaml#L2

**Comment by webvictim** [comment]
@pierreyves-lebrun Could you please add some tests to `deployment_test.yaml` which validate the functionality that you added?

**Comment by pierreyves-lebrun** [comment]
> @pierreyves-lebrun Could you please add some tests to `deployment_test.yaml` which validate the functionality that you added?

@webvictim Added a few tests, please review again

File: examples/chart/teleport-cluster/templates/deployment.yaml [hunk 3]
```diff
@@ -230,9 +231,9 @@ spec:
         configMap:
           name: {{ .Release.Name }}
       - name: "data"
-        {{- if eq .Values.chartMode "standalone" }}
+        {{- if or (eq .Values.chartMode "standalone") (eq .Values.chartMode "custom") }}
```
[the remark this answers is not in the document]
Comment (pierreyves-lebrun) [reply examples/chart/teleport-cluster/templates/deployment.yaml]: Added `persistence.enabled: true` as this is the convention used in most charts.

I am aware that this is a slightly breaking change for `chartMode: custom` users, but I don't think anybody will miss losing data upon upgrading the chart, which was an annoying "feature" 😅

**Comment by pierreyves-lebrun** [comment]
> /gcbrun

All jobs passed but I rebased onto master so you might want to rerun the pipeline

File: examples/chart/teleport-cluster/templates/deployment.yaml [hunk 4]
```diff
@@ -230,9 +231,9 @@ spec:
         configMap:
           name: {{ .Release.Name }}
       - name: "data"
-        {{- if eq .Values.chartMode "standalone" }}
+        {{- if and (.Values.persistence.enabled) (or (eq .Values.chartMode "standalone") (eq .Values.chartMode "custom")) }}
```
Comment (webvictim) [inline examples/chart/teleport-cluster/templates/deployment.yaml:234]: If I'm reading this right, I think we may now have a different backwards compatibility problem to before...

`custom` mode is now fine and can optionally use `persistence.enabled`, but if someone is using the legacy `standalone` mode (i.e. where storage previously always defaulted to enabled), then they won't be setting `persistence.enabled: true` and their current PVC (and thus Teleport state) would vanish when they upgrade the chart.

I think the way it needs to work is like this:
```yaml
chartMode: custom
persistence:
  enabled: true
# PVC enabled

chartMode: custom
# PVC disabled

chartMode: standalone
persistence:
  enabled: true
# PVC enabled

chartMode: standalone
# PVC still enabled
```

File: examples/chart/teleport-cluster/tests/deployment_test.yaml [hunk 5]
```diff
@@ -63,11 +63,11 @@ tests:
...
-            - matchExpressions:
-              - key: gravitational.io/dedicated
-                operator: In
-                values:
-                - teleport
+              - matchExpressions:
+                  - key: gravitational.io/dedicated
+                    operator: In
+                    values:
+                      - teleport
```
Comment (webvictim) [inline examples/chart/teleport-cluster/tests/deployment_test.yaml:70]: Unrelated change?

**Review by webvictim — COMMENTED** [review]
Thanks again @pierreyves-lebrun. Just realised another backwards compatibility situation and a couple of other comments, then we're good to go. This is a tricky PR and I really don't want to break anyone's current setup which is why I'm being such a stickler - apologies!

**Comment by pierreyves-lebrun** [comment]
> Thanks again @pierreyves-lebrun. Just realised another backwards compatibility situation and a couple of other comments, then we're good to go. This is a tricky PR and I really don't want to break anyone's current setup which is why I'm being such a stickler - apologies!

No worries, it's definitely worth being cautious when it comes to data persistence

File: examples/chart/teleport-cluster/templates/deployment.yaml (hunk 4 above)
Comment (pierreyves-lebrun) [reply examples/chart/teleport-cluster/templates/deployment.yaml:234]: Well caught! Will find a way to handle that

Comment (pierreyves-lebrun) [reply examples/chart/teleport-cluster/templates/deployment.yaml:234]: > Well caught! Will find a way to handle that

Actually I don't think there is an issue as the default `persistence.enabled: true` is always used by the chart, regardless of whether `standalone:` section is specified or not. The only way for `chartMode: standalone` users to lose data would be to either remove `persistence.enabled` field or set `persistence.enabled: false` from the chart original `values.yaml`.

Please note that directly editing/replacing a chart `values.yaml` (e.g.: pasting the content of an older chart `values.yaml` to the latest one) can be regarded as bad practise since default values might then be missing. `helm upgrade -f override.yaml ...` or `helm upgrade --set ...` is the correct and safest approach.

Of course don't hesitate to tell me if I misinterpreted your question or missed something, in fact your question made me realize another edge case related to persistenVolumeClaim size (fixed [here](https://github.com/gravitational/teleport/pull/11368/commits/371f557701c367a387fe1161ce71e32dcade1465)).

Comment (webvictim) [reply examples/chart/teleport-cluster/templates/deployment.yaml:234]: Got it. I agree with your assessment and would hope most people are using the chart correctly i.e. allowing the default values to apply, and then overriding individual specific values on the command line with `-f values.yaml`. 

If someone has incorrectly copy/pasted an older version of the entire default values file and are then setting each thing there deliberately, I don't think this is a use-case we should be expected to support. Even then, they should have no value set for `persistence` which will mean they default to `persistence.enabled: true` and everything will work anyway.

File: examples/chart/teleport-cluster/values.yaml [hunk 6]
```diff
@@ -74,10 +74,12 @@ labels: {}
...
 
-################################################################
-# Standalone-specific settings (only used in "standalone" mode)
-################################################################
-standalone:
+######################################################################
+# Persistence settings (only used in "standalone" and "custom" modes)
+######################################################################
+persistence:
+  ## Enable persistence using Persistent Volume Claims
```
Comment (webvictim) [inline examples/chart/teleport-cluster/values.yaml:81]:
```suggestion
  # Enable persistence using a PersistentVolumeClaim
```

**Review by webvictim — APPROVED** [review]
LGTM - just one final nitpick in the values file. We do need to add these new values to the Helm reference for the `teleport-cluster` chart, but given how much work you've done on the PR already, I'm OK with making those updates myself in a separate PR. Thanks @pierreyves-lebrun!

**Comment by pierreyves-lebrun** [comment]
> LGTM - just one final nitpick in the values file. We do need to add these new values to the Helm reference for the `teleport-cluster` chart, but given how much work you've done on the PR already, I'm OK with making those updates myself in a separate PR. Thanks @pierreyves-lebrun!

I had totally overlooked that, it's updated now so feel free to give me further feedback

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx [hunk 7]
```diff
@@ -407,39 +407,61 @@ Teleport's RBAC policies to define access rules for the cluster.
 | `gcp` | [Running an HA Teleport cluster using a Google Cloud GKE cluster](../guides/gcp.mdx) |
 | `custom` | [Running a Teleport cluster with a custom config](../guides/custom.mdx) |
 
-## `standalone`
+## `persistence`
 
-### `standalone.existingClaimName`
+### `[persistence].enabled`
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx:412]: I think if we make these a link it'll mess with the TOC/formatting.

```suggestion
### `persistence.enabled`
```

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx [hunk 8]
```diff
@@ -407,39 +407,61 @@ Teleport's RBAC policies to define access rules for the cluster.
...
+<Tabs>
+  <TabItem label="values.yaml">
+  ```yaml
+  persistence:
+    enabled: true
+  ```
+  </TabItem>
+  <TabItem label="--set">
+  ```code
+  $ --set [persistence].enabled=true
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx:429]:
```suggestion
  $ --set persistence.enabled=true
```

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx [hunk 9]
```diff
@@ -407,39 +407,61 @@ Teleport's RBAC policies to define access rules for the cluster.
...
+  ```
+  </TabItem>
+</Tabs>
+
+### `[persistence].existingClaimName`
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx:434]:
```suggestion
### `persistence.existingClaimName`
```

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx [hunk 10]
```diff
@@ -407,39 +407,61 @@ Teleport's RBAC policies to define access rules for the cluster.
...
 
 | Type | Default value | Can be used in `custom` mode? |
 | - | - | - |
 | `string` | `nil` | ✅ |
 
-`standalone.existingClaimName` can be used to provide the name of a pre-existing `PersistentVolumeClaim` to use if desired.
+`[persistence].existingClaimName` can be used to provide the name of a pre-existing `PersistentVolumeClaim` to use if desired.
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx:440]:
```suggestion
`persistence.existingClaimName` can be used to provide the name of a pre-existing `PersistentVolumeClaim` to use if desired.
```

File: docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx [hunk 11]
```diff
@@ -407,39 +407,61 @@ Teleport's RBAC policies to define access rules for the cluster.
...
   <TabItem label="--set">
   ```code
-  $ --set standalone.existingClaimName=my-existing-pvc-name
+  $ --set [persistence].existingClaimName=my-existing-pvc-name
   ```
   </TabItem>
 </Tabs>
 
-### `standalone.volumeSize`
+### `[persistence].volumeSize`
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx:458]:
```suggestion
### `persistence.volumeSize`
```

**Review by webvictim — COMMENTED** [review]
Amazing, thank you!

**Comment by pierreyves-lebrun** [comment]
> Amazing, thank you!

Cheers, all suggested changes should be in

**Review by webvictim — COMMENTED** [review]
Couple more I noticed

**Review by zmb3 — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Comment by webvictim** [comment]
Hey @pierreyves-lebrun, thanks for all the work. I've opened #11693 and will merge this there - our Github bot sometimes always handle PRs from forks very well (and this is one of those cases)

**Comment by stefansedich** [comment]
> Thanks again @pierreyves-lebrun. Just realised another backwards compatibility situation and a couple of other comments, then we're good to go. This is a tricky PR and I really don't want to break anyone's current setup which is why I'm being such a stickler - apologies!

This one broke out set 😆 we don't use persistence and found after a 9.x upgrade it was turned on by default, as we run multiple replicas things went 💥

**Comment by webvictim** [comment]
Sorry @stefansedich. I think that enabling persistence by default is what we do want for the chart, just to avoid people failing to set it up and then inadvertently losing cluster data when upgrading.

I apologise that it broke your setup - hopefully setting `persistance.enabled: false` in your values ended up fixing things.
